lighthouse-core/gather/gatherers/source-maps.js

```diff
@@ -0,0 +1,118 @@
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+const Driver = require('../driver.js'); // eslint-disable-line no-unused-vars
+
+/* global fetch */
+
+/**
+ * This function is careful not to parse the response as JSON, as it will
+ * just need to be serialized again over the protocol, and source maps can
+ * be huge.
+ *
+ * @param {string} url
+ */
+/* istanbul ignore next */
+async function fetchSourceMap(url) {
+  try {
+    const response = await fetch(url);
+    return response.text();
+  } catch (err) {
+    return {errorMessage: err.toString()};
+  }
+}
+
+/**
+ * @fileoverview Gets JavaScript source maps.
+ */
+class SourceMaps extends Gatherer {
+  constructor() {
+    super();
+    /** @type {LH.Crdp.Debugger.ScriptParsedEvent[]} */
+    this._scriptParsedEvents = [];
+    this.onScriptParsed = this.onScriptParsed.bind(this);
+  }
+
+  /**
+   * @param {Driver} driver
+   * @param {string} sourceMapUrl
+   * @return {Promise<{map: import('source-map').RawSourceMap} | {errorMessage: string}>}
+   */
+  async fetchSourceMapInPage(driver, sourceMapUrl) {
+    // TODO: change default protocol timeout?
+    // driver.setNextProtocolTimeout(250);
+    /** @type {string|{errorMessage: string}} */
+    const sourceMapJson =
+      await driver.evaluateAsync(`(${fetchSourceMap})(${JSON.stringify(sourceMapUrl)})`);
+
+    if (typeof sourceMapJson === 'object') {
+      return sourceMapJson;
+    }
+
+    return {map: JSON.parse(sourceMapJson)};
+  }
+
+  /**
+   * @param {LH.Crdp.Debugger.ScriptParsedEvent} event
+   */
+  onScriptParsed(event) {
+    if (event.sourceMapURL) {
+      this._scriptParsedEvents.push(event);
+    }
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   */
+  async beforePass(passContext) {
+    const driver = passContext.driver;
+    driver.on('Debugger.scriptParsed', this.onScriptParsed);
+    await driver.sendCommand('Debugger.enable');
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts['SourceMaps']>}
+   */
+  async afterPass(passContext) {
+    const driver = passContext.driver;
+
+    driver.off('Debugger.scriptParsed', this.onScriptParsed);
+    await driver.sendCommand('Debugger.disable');
+
+    /** @type {LH.Artifacts.SourceMap[]} */
+    const sourceMaps = [];
+    for (const event of this._scriptParsedEvents) {
+      if (!event.sourceMapURL) continue;
+
+      const url = event.url;
+      try {
+        if (event.sourceMapURL.startsWith('data:')) {
+          const buffer = Buffer.from(event.sourceMapURL.split(',')[1], 'base64');
+          sourceMaps.push({
+            url,
+            map: JSON.parse(buffer.toString()),
+          });
+        } else {
+          const fetchedSourceMap = await this.fetchSourceMapInPage(driver, event.sourceMapURL);
```

Comment:
if we're fetching these serially then we should definitely have the protocol timeout fairly short.

is there a particular reason we're not doing them in parallel? might cleanup the async `.push` confusion too if we did a `Promise.all`

Reply:
if we do it in parallel, and make the timeout very short, wouldn't the latter ones (say there's like 15 maps) timeout b/c only a few requests get handled at a time?

Reply:
@brendankenny 	@patrickhulce 	 thoughts?

Reply:
from my perspective if we can't download all of them in a ~5-10s span then we should probably seriously reconsider trying to download all the source maps on every run :/

if we can then I see no reason not to just fire em all off with our max timeout and whatever times out, times out.